[WIP]Offset auto reset

lnbest0707-uber opened the pull request:
Instructions:
1. The PR has to be tagged with at least one of the following labels (*):
   1. `feature`
   2. `bugfix`
   3. `performance`
   4. `ui`
   5. `backward-incompat`
   6. `release-notes` (**)
2. Remove these instructions before publishing the PR.
 
(*) Other labels to consider:
- `testing`
- `dependencies`
- `docker`
- `kubernetes`
- `observability`
- `security`
- `code-style`
- `extension-point`
- `refactor`
- `cleanup`

(**) Use `release-notes` label for scenarios like:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed

deemoliu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/LLCSegmentName.java:
```diff
@@ -38,25 +38,59 @@ public class LLCSegmentName implements Comparable<LLCSegmentName> {
     String[] parts = StringUtils.splitByWholeSeparator(segmentName, SEPARATOR);
-    Preconditions.checkArgument(parts.length == 4, "Invalid LLC segment name: %s", segmentName);
+    Preconditions.checkArgument(
+        parts.length >= 4 && parts.length <= 5, "Invalid LLC segment name: %s", segmentName);
```
can you add an example in the comment for LLCSegmentName

lnbest0707-uber replied:
The example could be found in the test `pinot-common/src/test/java/org/apache/pinot/common/utils/LLCSegmentNameTest.java`
This file was without examples, but I can add in next patch

deemoliu commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/RealtimeOffsetAutoResetKafkaHandler.java:
```diff
@@ -0,0 +1,135 @@
+
+  private void removeTopicFromTableConfig(String tableNameWithType, String topicName, TableConfig tableConfig) {
+    List<Map<String, String>> streamConfigMaps = IngestionConfigUtils.getStreamConfigMaps(tableConfig);
+    for (int i = 0; i < streamConfigMaps.size(); i++) {
```
shall we remove from index n-1 to 0 which is more robust?

deemoliu replied:
Major: 
Q1: Does streamConfigMaps maintain the order?
Q2: I feel it might be a concurrent modification exception here.

deemoliu commented on pinot-spi/src/main/java/org/apache/pinot/spi/stream/PartitionGroupConsumptionStatus.java:
```diff
@@ -35,15 +35,18 @@
  */
 public class PartitionGroupConsumptionStatus {
+
+  private final String _topicName;
```
topicName seems to be a notNull value, let's verify this carefully in the integration tests.

deemoliu commented on pinot-spi/src/main/java/org/apache/pinot/spi/utils/IngestionConfigUtils.java:
```diff
@@ -124,6 +126,25 @@ public static int getStreamConfigIndexFromPinotPartitionId(int partitionId) {
+      int partitionId, String topicName, List<StreamConfig> streamConfigs) {
+    int rawTopicIndex = getStreamConfigIndexFromPinotPartitionId(partitionId);
+    if (topicName != null && !topicName.isEmpty()) {
+      return streamConfigs.stream().filter(c -> topicName.equals(c.getTopicName()))
```
If multiple StreamConfigs match a same topic name, only the first is returned.
we should clarify we don't support multiple "same topic different kafka cluster" in the streamConfigs.

deemoliu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeOffsetAutoResetManager.java:
```diff
@@ -0,0 +1,240 @@
+  protected RealtimeOffsetAutoResetManager.Context preprocess(Properties periodicTaskProperties) {
+    RealtimeOffsetAutoResetManager.Context context = new RealtimeOffsetAutoResetManager.Context();
+    // Fill offset back fill job required info
+    if (periodicTaskProperties.containsKey(context._backfillJobPropertyKeys.toArray()[0])) {
```
can you add pointers for a test or a comment/javaDoc for an example of periodTaskProperties.

deemoliu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeOffsetAutoResetManager.java:
```diff
@@ -0,0 +1,240 @@
+      _tableToHandler.get(tableNameWithType).triggerBackfillJob(tableNameWithType,
+          topicStreamConfig,
+          topicName,
+          Integer.valueOf(context._backfillJobProperties.get(Constants.RESET_OFFSET_TOPIC_PARTITION)),
```
try catch numberFormatException?

deemoliu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeOffsetAutoResetManager.java:
```diff
@@ -0,0 +1,240 @@
+
+  @Override
+  protected void nonLeaderCleanup(List<String> tableNamesWithType) {
+    for (String tableNameWithType : tableNamesWithType) {
```
do we need to cleanup _tableEphemeralTopics in this setup

deemoliu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeOffsetAutoResetManager.java:
```diff
@@ -0,0 +1,240 @@
+    RealtimeOffsetAutoResetHandler handler = getOrConstructHandler(tableConfig);
+    Collection<String> cleanedUpTopics = handler.cleanupCompletedBackfillJobs(
+        tableNameWithType, _tableEphemeralTopics.get(tableNameWithType));
+    if (cleanedUpTopics.size() >= _tableEphemeralTopics.get(tableNameWithType).size()) {
```
suggest: 
if (cleanedUpTopics.containsAll(_tableEphemeralTopics.get(tableNameWithType))) {
...
}

if that possible we have uncleaned topics from the previous round?

lnbest0707-uber replied:
revised

deemoliu reviewed (commented):
to my curiosity, are there any place we can tune the periodically job frequency, thanks!

lnbest0707-uber commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/RealtimeOffsetAutoResetKafkaHandler.java:
```diff
@@ -0,0 +1,135 @@
+      LOGGER.error("Cannot remove backfill topics {} from the table config", topicNames, e);
+      cleanedUpTopics.clear();
+    }
+    return cleanedUpTopics;
```
It would periodically check and remove if the **ephemeral** topic is **complete**.

lnbest0707-uber commented:
> please address the conflict and rerun the intergation tests, thank you!

@deemoliu  Thanks for the review. Addressed the comments. The failing test looks to be flaky. Rerun twice and failed in different locations (the failed in 1st one succeed in 2nd run, but the 2nd run failed at the place where it succeeded in 1st)

lnbest0707-uber commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/RealtimeOffsetAutoResetKafkaHandler.java:
```diff
@@ -0,0 +1,135 @@
+import org.slf4j.LoggerFactory;
+
+
+public abstract class RealtimeOffsetAutoResetKafkaHandler extends RealtimeOffsetAutoResetHandler {
```
We are internally using a Kafka replication based implementation (refer to the attached design doc). That one uses the Kafka ecosystem APIs to trigger the backfill. That particular implementation is pretty "internal". That's the reason I make it pluggable here.

lnbest0707-uber commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/RealtimeOffsetAutoResetKafkaHandler.java:
```diff
@@ -0,0 +1,135 @@
+    }
+    try {
+      TableConfig currentTableConfig = _pinotHelixResourceManager.getTableConfig(tableNameWithType);
+      addNewTopicToTableConfig(newTopicStreamConfig, currentTableConfig);
```
Yes it is. The design doc explains it. I would also add comments.

lnbest0707-uber commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeOffsetAutoResetManager.java:
```diff
@@ -0,0 +1,248 @@
+    //   resetOffsetTopicPartition=0
+    //   resetOffsetFrom=0
+    //   resetOffsetTo=1000
+    if (periodicTaskProperties.containsKey(context._backfillJobPropertyKeys.toArray()[0])) {
```
> For periodic task, periodicTaskProperties applies to all tables. How do you use it to setup the topic info for one single partition?
Not sure if I understand correctly.
When the task is triggered by the "periodic", the `periodicTaskProperties` is empty and the manager/handler would try to check if the registered backfill job is running or complete. If complete, it would remove the backfill topic.
When the task is triggered by 'PinotLLCRealtimeSegmentManager` during segment commit, then `periodicTaskProperties` would have required info to "trigger" the backfill.

Jackie-Jiang replied:
Do you want to trigger this manually only? From periodic trigger, `context._shouldTriggerBackfillJobs` will be `false`, and the task becomes no-op. We shouldn't even scan tables

lnbest0707-uber commented on pinot-common/src/main/java/org/apache/pinot/common/utils/LLCSegmentName.java:
```diff
@@ -38,25 +38,62 @@ public class LLCSegmentName implements Comparable<LLCSegmentName> {
+    this(tableName, "", partitionGroupId, sequenceNumber, msSinceEpoch);
+  }
+
+  public LLCSegmentName(
```
This is a good point and it requires reverse update sequence (if this introduced feature is already in use).

The update is useful even for normal multi-topic ingestions. With that, we will be able to remove topics from existing table. We've seen a lot of such request from our running production.

Please also suggest any better or more graceful methods. Thanks

Jackie-Jiang replied:
This is a good change, but needs to follow one of the 2 ways:
- Introduce it in 2 releases
  - First release add all handling of `topic` with backward compatibility to old segment name but not change the segment name
  - Second release change the segment name after all nodes can already handle it
- Add a flag to control the segment name format (e.g. one cluster config + per table override in table config), and by default using the existing name format

I prefer the second one for faster rolling out + not changing existing behavior. Making segment name longer has side effect of larger ZNode size, and will cause Pinot to hit scale limit of ZK faster. For single topic case we might still want to stick with the current format.

Given this change itself is very critical, we should split it out as a separate PR

lnbest0707-uber commented on pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:
```diff
@@ -275,6 +279,7 @@ public static long getRandomInitialDelayInSeconds() {
     public static final int DEFAULT_RETENTION_MANAGER_FREQUENCY_IN_SECONDS = 6 * 60 * 60; // 6 Hours.
     public static final int DEFAULT_OFFLINE_SEGMENT_INTERVAL_CHECKER_FREQUENCY_IN_SECONDS = 24 * 60 * 60; // 24 Hours.
     public static final int DEFAULT_REALTIME_SEGMENT_VALIDATION_FREQUENCY_IN_SECONDS = 60 * 60; // 1 Hour.
+    public static final int DEFAULT_OFFSET_AUTO_RESET_FREQUENCY_IN_SECONDS = 60 * 60; // 1 Hour.
```
Good point, I can make the manager to be enabled by a config.

lnbest0707-uber commented:
> Can we split the PR into 2 parts:
> 
> 1. Automatically reset offset to largest offset
> 2. Handle backfill
> 
> @KKcorps @noob-se7en Please also take a look

Thanks for the review.
Wanna clarify about the "split", the PR itself is right now with only interfaces for backfill. The change that "matters" in other perspective is to introduce the "backfill" type of topic (different naming method). Do you suggest making that part independent or making the new manager independent?
